mtr.8
.TH MTR 8 "March 4, 1999" "mtr" "mtr"


.SH NAME
mtr \- a network diagnostic tool


.SH SYNOPSIS
.B mtr 
[\c
.B \-hvrctglspeniu46\c
]
[\c
.B \-\-help\c
]
[\c
.B \-\-version\c
]
[\c
.B \-\-report\c
]
[\c
.B \-\-report-wide\c
]
[\c
.B \-\-report\-cycles\ COUNT\c
]
[\c
.B \-\-curses\c
]
[\c
.B \-\-split\c
]
[\c
.B \-\-raw\c
]
[\c
.B \-\-mpls\c
]
[\c
.B \-\-no-dns\c
]
[\c
.B \-\-gtk\c
]
[\c
.B \-\-address\ IP.ADD.RE.SS\c
]
[\c
.B \-\-interval\ SECONDS\c
]
[\c
.B \-\-psize\ BYTES | -s BYTES\c
]
.B HOSTNAME [PACKETSIZE]


.SH DESCRIPTION

.B mtr 
combines the functionality of the 
.B traceroute
and 
.B ping
programs in a single network diagnostic tool.

.PP
As 
.B mtr 
starts, it investigates the network connection between the host 
.B mtr
runs on and 
.BR HOSTNAME . 
by sending packets with purposly low TTLs. It continues to send
packets with low TTL, noting the response time of the intervening
routers.  This allows 
.B mtr 
to print the response percentage and response times of the internet
route to 
.BR HOSTNAME . 
A sudden increase in packetloss or response time is often an indication
of a bad (or simply overloaded) link. 

.SH OPTIONS

.TP
.B \-h
.TP
.B \-\-help
.br
Print the summary of command line argument options.

.TP
.B \-v
.TP
.B \-\-version
.br
Print the installed version of mtr.  

.TP
.B \-r
.TP
.B \-\-report
.br
This option puts 
.B mtr
into 
.B report
mode.  When in this mode,
.B mtr
will run for the number of cycles specified by the 
.B \-c
option, and then print statistics and exit.  
.TP
\c
This mode is useful for generating statistics about network quality.  
Note that each running instance of 
.B mtr
generates a significant amount of network traffic.  Using 
.B mtr
to measure the quality of your network may result in decreased
network performance.  

.TP
.B \-w
.TP
.B \-\-report-wide
.br
This option puts 
.B mtr
into 
.B wide report
mode.  When in this mode,
.B mtr
will not cut hostnames in the report. 

.TP
.B \-c\ COUNT
.TP
.B \-\-report\-cycles\ COUNT
Use this option to set the number of pings sent to determine
both the machines on the network and the reliability of 
those machines.  Each cycle lasts one second.

.TP
.B \-s\ BYTES
.TP
.B \-\-psize\ BYTES
.TP
.B PACKETSIZE
These options or a trailing PACKETSIZE on the commandline sets 
the packet size used for probing.
It is in bytes inclusive IP and ICMP headers

If set to a negative number, every iteration will use a different, random
packetsize upto that number. 
.TP
.B \-t
.TP
.B \-\-curses
.br
Use this option to force 
.B mtr 
to use the curses based terminal
interface (if available).

.TP
.B \-e
.TP
.B \-\-mpls
.br
Use this option to tell 
.B mtr 
to display information from ICMP extensions for MPLS (RFC 4950)
that are encoded in the response packets.

.TP
.B \-n
.TP
.B \-\-no-dns
.br
Use this option to force 
.B mtr 
to display numeric IP numbers and not try to resolve the
host names. 

.TP
.B \-o\ fields\ order
.TP
.B \-\-order\ fields\ order
.br
Use this option to specify the fields and their order when loading mtr.
.br
Example:
-o "LSD NBAW"

.TP
.B \-g
.TP
.B \-\-gtk
.br
Use this option to force
.B mtr 
to use the GTK+ based X11 window interface (if available).  
GTK+ must have been available on the system when 
.B mtr 
was built for this to work.  See the GTK+ web page at 
.B http://www.gtk.org/
for more information about GTK+.

.TP
.B \-p
.TP
.B \-\-split
.br
Use this option to set
.B mtr 
to spit out a format that is suitable for a split-user interface.

.TP
.B \-l
.TP
.B \-\-raw
.br
Use this option to tell 
.B mtr 
to use the raw output format. This format is better suited for
archival of the measurement results. It could be parsed to 
be presented into any of the other display methods. 

.TP
.B \-a\ IP.ADD.RE.SS
.TP
.B \-\-address\ IP.ADD.RE.SS
.br
Use this option to bind outgoing packets' socket to specific interface,
so that any packet will be sent through this interface. NOTE that this
option doesn't apply to DNS requests (which could be and could not be 
what you want).

.TP
.B \-i\ SECONDS
.TP
.B \-\-interval\ SECONDS
.br
Use this option to specify the positive number of seconds between ICMP
ECHO requests.  The default value for this parameter is one second.

.TP
.B \-u
.br
Use UDP datagrams instead of ICMP ECHO.

.TP
.B \-4
.br
Use IPv4 only.

.TP
.B \-6
.br
Use IPv6 only.

.SH BUGS

Some modern routers give a lower priority to ICMP ECHO packets than 
to other network traffic.  Consequently, the reliability of these
routers reported by 
.B mtr
will be significantly lower than the actual reliability of 
these routers.  


.SH CONTACT INFORMATION

.PP
For the latest version, see the mtr web page at 
.BR http://www.bitwizard.nl/mtr/ .

.PP
Subscribe to the mtr mailing list.  All mtr related announcements
are posted to the mtr mailing list.  To subscribe, send email to
.B majordomo@lists.xmission.com
with 
.B subscribe\ mtr
in the body of the message.  To send a message to the mailing list, mail to 
.BR mtr@lists.xmission.com .

.PP
Bug reports and feature requests should be sent to the mtr
mailing list.


.SH "SEE ALSO"

traceroute(8),
ping(8).


debian/changelog
mtr (0.82-3ubuntu1) quantal; urgency=low

  * Merge from Debian Unstable (LP: #1014017). Remaining Changes:
    - Apply man page patch from Daniël van Eeden updating the GTK+
      web site URL. (LP: #817429)
 
 -- Vibhav Pant <vibhavp@gmail.com>  Sat, 16 Jun 2012 15:45:41 +0530

mtr (0.82-3) unstable; urgency=low

  * Added patch by Raúl Benencia to fix FTBFS by generating aclocal.m4
    before compiling, closes: #662583

 -- Robert Woodcock <rcw@debian.org>  Sun, 15 Apr 2012 14:40:05 -0700

mtr (0.82-2) unstable; urgency=low

  * Add patch by Moritz Mühlenhoff to enable hardened build flags,
    closes: #654117.

 -- Robert Woodcock <rcw@debian.org>  Sun, 08 Jan 2012 08:05:30 -0800

mtr (0.82-1) unstable; urgency=low

  * New upstream version fixes FTBFS, closes: #650535

 -- Robert Woodcock <rcw@debian.org>  Sun, 25 Dec 2011 09:40:19 -0800

mtr (0.81-1) unstable; urgency=low

  * New upstream version

 -- Robert Woodcock <rcw@debian.org>  Sat, 26 Nov 2011 20:15:42 -0800

mtr (0.80-1) unstable; urgency=low

  * New upstream version
  * Add Homepage field to debian/control, closes: #314651
  * Change Priority field to match that of the archive overrides,
    closes: #416567
  * Lintian cleanups

 -- Robert Woodcock <rcw@debian.org>  Mon, 06 Sep 2010 16:53:54 -0700

mtr (0.75-2) unstable; urgency=low

  * Use rm -f in rules clean target where necessary, closes: #499789
 
 -- Robert Woodcock <rcw@debian.org>  Mon, 22 Sep 2008 07:30:21 -0700

mtr (0.75-1) unstable; urgency=low

  * New upstream version with rewritten GTK+ code
    closes: #65005, #30263
  * Bumped Standards-Version to 3.8.0
  * Lintian cleanups

 -- Robert Woodcock <rcw@debian.org>  Sun, 21 Sep 2008 10:39:10 -0700

mtr (0.73-1) unstable; urgency=low

  * New upstream version

 -- Robert Woodcock <rcw@debian.org>  Wed, 16 Apr 2008 19:38:12 -0700

mtr (0.72-2) unstable; urgency=low

  * Reupload to unstable since etch is released.
  * Include no-busy-wait patch in #156378,
    closes: #156378, #259906, #348650, #384444
  * Use debhelper compatibility level 5

 -- Robert Woodcock <rcw@debian.org>  Fri, 25 May 2007 21:42:04 -0700

mtr (0.72-1) experimental; urgency=low

  * New upstream version fixes IPv6 reverse DNS lookups for 8-bit
    hostnames, closes: #384926

 -- Robert Woodcock <rcw@debian.org>  Mon, 01 Jan 2007 15:33:33 -0800

mtr (0.71-2) unstable; urgency=low

  * Run autoreconf --force --install to allow building without automake,
    closes: #365845

 -- Robert Woodcock <rcw@debian.org>  Sun, 14 May 2006 09:37:17 -0700

mtr (0.71-1) unstable; urgency=low

  * New upstream version
  * Follow autotools-dev suggestions regarding timestamping, closes: #362614

 -- Robert Woodcock <rcw@debian.org>  Tue, 02 May 2006 18:24:05 -0700

mtr (0.70-1) unstable; urgency=low

  * New upstream version

 -- Robert Woodcock <rcw@debian.org>  Thu, 23 Mar 2006 17:31:28 -0800

mtr (0.69-4) unstable; urgency=low

  * Woops, forgot to run autoconf for the configure.in change in -3,
    closes: #254089

 -- Robert Woodcock <rcw@debian.org>  Sun, 05 Mar 2006 18:19:05 -0800

mtr (0.69-3) unstable; urgency=low

  * Added check for __res_mkquery to build on amd64,
    closes: #166613, #254089, #257981, #354342
  * Uppercased initial letter of menu title, closes: #320609
  * Fixed manual page and PrintHelp function for -p/-s switcharoo,
    closes: #305728
  * DNS fix from Jarek Kaminski if no nameservers are configured,
    closes: #301346

 -- Robert Woodcock <rcw@debian.org>  Fri, 03 Mar 2006 15:16:27 -0800

mtr (0.69-2) unstable; urgency=low

  * Added upstream source location to copyright file, closes #312670

 -- Robert Woodcock <rcw@debian.org>  Fri, 10 Jun 2005 19:29:46 -0700

mtr (0.69-1) unstable; urgency=low

  * New upstream version adds IPv6 support
  * Fixed copyright file, closes #290215
  * Allow for IPv4 operation on IPv4-only systems
  * Copy addr buffer in net_process_ping immediately to avoid overwriting it

 -- Robert Woodcock <rcw@debian.org>  Wed, 23 Mar 2005 19:39:59 -0800

mtr (0.67-1) unstable; urgency=high

  * SECURITY: New upstream version fixes off-by-one error with MAXFLD,
    closes: #285228

 -- Robert Woodcock <rcw@debian.org>  Sun, 12 Dec 2004 11:26:07 -0800

mtr (0.63-1.2) unstable; urgency=low

  * Non-maintainer upload.
  * Apply patch from Christoph Berg <cb@df7cb.de> converting fld_index[] from
    char to int in order to work around the invalid assumption that plain char++
    is always unsigned (closes: #283756).

 -- Mark Brown <broonie@debian.org>  Thu,  9 Dec 2004 11:39:42 +0000

mtr (0.63-1.1) unstable; urgency=low

  * NMU.
  * Remove useless dependency on non-existing automake-1.5 and
    autoconf (Closes: #266006).

 -- Christoph Berg <cb@df7cb.de>  Sat, 27 Nov 2004 01:34:28 +0100

mtr (0.63-1) unstable; urgency=low

  * New upstream version
  * New version compiles against gtk2.4, closes: #262606
  * New version validates input to order-of-fields option, preventing
    segfaults, closes: #264055, #264694
  * Fixed virtual dependancy on automaken in response to #257981

 -- Robert Woodcock <rcw@debian.org>  Sun, 15 Aug 2004 16:35:32 -0700

mtr (0.58-1) unstable; urgency=low

  * New upstream version, closes: #252369

 -- Robert Woodcock <rcw@debian.org>  Sat, 03 Jul 2004 13:59:24 -0700

mtr (0.54-1) unstable; urgency=low

  * New upstream version

 -- Robert Woodcock <rcw@debian.org>  Sat,  3 May 2003 08:57:48 -0700

mtr (0.53-1) unstable; urgency=low

  * New upstream version, closes: #189692

 -- Robert Woodcock <rcw@debian.org>  Wed, 23 Apr 2003 19:08:15 -0700

mtr (0.52-1) unstable; urgency=low

  * New upstream version

 -- Robert Woodcock <rcw@debian.org>  Sun, 24 Nov 2002 12:11:51 -0800

mtr (0.51-1) unstable; urgency=low

  * New upstream version fixes hanging --report option, closes: #148045
  * Reverted some makefile reversions in 0.51

 -- Robert Woodcock <rcw@debian.org>  Fri, 24 May 2002 16:06:34 -0700

mtr (0.50-1) unstable; urgency=low

  * New upstream version

mtr.c
/*
    mtr  --  a network diagnostic tool
    Copyright (C) 1997,1998  Matt Kimball

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License version 2 as 
    published by the Free Software Foundation.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include <sys/types.h>
#include <config.h>
#include <netdb.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <netinet/in.h>
#include <sys/socket.h> 
#include <unistd.h>
#include <strings.h>

#include "mtr.h"
#include "mtr-curses.h"
#include "getopt.h"
#include "display.h"
#include "dns.h"
#include "report.h"
#include "net.h"


#ifdef ENABLE_IPV6
#define DEFAULT_AF AF_UNSPEC
#else
#define DEFAULT_AF AF_INET
#endif


#ifdef NO_HERROR
#define herror(str) fprintf(stderr, str ": error looking up \"%s\"\n", Hostname);
#endif


int   DisplayMode;
int   display_mode;
int   Interactive = 1;
int   PrintVersion = 0;
int   PrintHelp = 0;
int   MaxPing = 10;
int   ForceMaxPing = 0;
float WaitTime = 1.0;
char *Hostname = NULL;
char *InterfaceAddress = NULL;
char  LocalHostname[128];
int   dns = 1;
int   enablempls = 0;
int   cpacketsize = 64;          /* default packet size */
int   bitpattern = 0;
int   tos = 0;
int   reportwide = 0;
int af = DEFAULT_AF;
int mtrtype = IPPROTO_ICMP;     /* Use ICMP as default packet type */

                                /* begin ttl windows addByMin */
int  fstTTL = 1;                /* default start at first hop */
/*int maxTTL = MaxHost-1;  */     /* max you can go is 255 hops */
int   maxTTL = 30;              /* inline with traceroute */
                                /* end ttl window stuff. */
int remoteport = 80;            /* for TCP tracing */
int timeout = 10 * 1000000;     /* for TCP tracing */


/* default display field(defined by key in net.h) and order */
unsigned char fld_active[2*MAXFLD] = "LS NABWV";
int           fld_index[256];
char          available_options[MAXFLD];


struct fields data_fields[MAXFLD] = {
  /* key, Remark, Header, Format, Width, CallBackFunc */
  {' ', "<sp>: Space between fields", " ",  " ",        1, &net_drop  },
  {'L', "L: Loss Ratio",          "Loss%",  " %4.1f%%", 6, &net_loss  },
  {'D', "D: Dropped Packets",     "Drop",   " %4d",     5, &net_drop  },
  {'R', "R: Received Packets",    "Rcv",    " %5d",     6, &net_returned},
  {'S', "S: Sent Packets",        "Snt",    " %5d",     6, &net_xmit  },
  {'N', "N: Newest RTT(ms)",      "Last",   " %5.1f",   6, &net_last  },
  {'B', "B: Min/Best RTT(ms)",    "Best",   " %5.1f",   6, &net_best  },
  {'A', "A: Average RTT(ms)",     "Avg",    " %5.1f",   6, &net_avg   },
  {'W', "W: Max/Worst RTT(ms)",   "Wrst",   " %5.1f",   6, &net_worst },
  {'V', "V: Standard Deviation",  "StDev",  " %5.1f",   6, &net_stdev },
  {'G', "G: Geometric Mean",      "Gmean",  " %5.1f",   6, &net_gmean },
  {'J', "J: Current Jitter",      "Jttr",   " %4.1f",   5, &net_jitter},
  {'M', "M: Jitter Mean/Avg.",    "Javg",   " %4.1f",   5, &net_javg  },
  {'X', "X: Worst Jitter",        "Jmax",   " %4.1f",   5, &net_jworst},
  {'I', "I: Interarrival Jitter", "Jint",   " %4.1f",   5, &net_jinta },
  {'\0', NULL, NULL, NULL, 0, NULL}
};


void init_fld_options (void)
{
  int i;

  for (i=0;i < 256;i++)
    fld_index[i] = -1;

  for (i=0;data_fields[i].key != 0;i++) {
    available_options[i] = data_fields[i].key;
    fld_index[data_fields[i].key] = i;
  }
  available_options[i] = 0;
}


void parse_arg (int argc, char **argv) 
{
  int opt;
  int i;
  static struct option long_options[] = {
    { "version", 0, 0, 'v' },
    { "help", 0, 0, 'h' },

    { "report", 0, 0, 'r' },
    { "report-wide", 0, 0, 'w' },
    { "xml", 0, 0, 'x' },
    { "curses", 0, 0, 't' },
    { "gtk", 0, 0, 'g' },
    { "raw", 0, 0, 'l' },
    { "split", 0, 0, 'p' },     /* BL */
    				/* maybe above should change to -d 'x' */

    { "order", 1, 0, 'o' },	/* fileds to display & their order */

    { "interval", 1, 0, 'i' },
    { "report-cycles", 1, 0, 'c' },
    { "psize", 1, 0, 's' },	/* changed 'p' to 's' to match ping option
				   overload psize<0, ->rand(min,max) */
    { "bitpattern", 1, 0, 'b' },/* overload b>255, ->rand(0,255) */
    { "tos", 1, 0, 'Q' },	/* typeof service (0,255) */
    { "mpls", 0, 0, 'e' },
    { "no-dns", 0, 0, 'n' },
    { "address", 1, 0, 'a' },
    { "first-ttl", 1, 0, 'f' },	/* -f & -m are borrowed from traceroute */
    { "max-ttl", 1, 0, 'm' },
    { "udp", 0, 0, 'u' },	/* UDP (default is ICMP) */
    { "tcp", 0, 0, 'T' },	/* TCP (default is ICMP) */
    { "port", 1, 0, 'P' },      /* target port number for TCP */
    { "timeout", 1, 0, 'z' },   /* timeout for TCP sockets */
    { "inet", 0, 0, '4' },	/* IPv4 only */
    { "inet6", 0, 0, '6' },	/* IPv6 only */
    { 0, 0, 0, 0 }
  };

  opt = 0;
  while(1) {
    /* added f:m:o: byMin */
    opt = getopt_long(argc, argv,
		      "vhrwxtglpo:i:c:s:b:Q:ena:f:m:uTP:z46", long_options, NULL);
    if(opt == -1)
      break;

    switch(opt) {
    case 'v':
      PrintVersion = 1;
      break;
    case 'h':
      PrintHelp = 1;
      break;

    case 'r':
      DisplayMode = DisplayReport;
      break;
    case 'w':
      reportwide = 1;
      break;
    case 't':
      DisplayMode = DisplayCurses;
      break;
    case 'g':
      DisplayMode = DisplayGTK;
      break;
    case 'p':                 /* BL */
      DisplayMode = DisplaySplit;
      break;
    case 'l':
      DisplayMode = DisplayRaw;
      break;
    case 'x':
      DisplayMode = DisplayXML;
      break;

    case 'c':
      MaxPing = atoi (optarg);
      ForceMaxPing = 1;
      break;
    case 's':
      cpacketsize = atoi (optarg);
      break;
    case 'a':
      InterfaceAddress = optarg;
      break;
    case 'e':
      enablempls = 1;
      break;
    case 'n':
      dns = 0;
      break;
    case 'i':
      WaitTime = atof (optarg);
      if (WaitTime <= 0.0) {
	fprintf (stderr, "mtr: wait time must be positive\n");
	exit (1);
      }
      if (getuid() != 0 && WaitTime < 1.0)
	WaitTime = 1.0;
      break;
    case 'f':
      fstTTL = atoi (optarg);
      if (fstTTL > maxTTL) {
	fstTTL = maxTTL;
      }
      if (fstTTL < 1) {                       /* prevent 0 hop */
	fstTTL = 1;
      }
      break;
    case 'm':
      maxTTL = atoi (optarg);
      if (maxTTL > (MaxHost - 1)) {
	maxTTL = MaxHost-1;
      }
      if (maxTTL < 1) {                       /* prevent 0 hop */
	maxTTL = 1;
      }
      if (fstTTL > maxTTL) {         /* don't know the pos of -m or -f */
	fstTTL = maxTTL;
      }
      break;
    case 'o':
      /* Check option before passing it on to fld_active. */
      if (strlen (optarg) > MAXFLD) {
	fprintf (stderr, "Too many fields: %s\n", optarg);
        exit (1);
      }
      for (i=0; optarg[i]; i++) {
        if(!strchr (available_options, optarg[i])) {
          fprintf (stderr, "Unknown field identifier: %c\n", optarg[i]);
          exit (1);
        }
      }
      strcpy ((char*)fld_active, optarg);
      break;
    case 'b':
      bitpattern = atoi (optarg);
      if (bitpattern > 255)
	bitpattern = -1;
      break;
    case 'Q':
      tos = atoi (optarg);
      if (tos > 255 || tos < 0) {
	/* error message, should do more checking for valid values,
	 * details in rfc2474 */
	tos = 0;
      }
      break;
    case 'u':
      if (mtrtype != IPPROTO_ICMP) {
        fprintf(stderr, "-u and -T are mutually exclusive.\n");
        exit(EXIT_FAILURE);
      }
      mtrtype = IPPROTO_UDP;
      break;
    case 'T':
      if (mtrtype != IPPROTO_ICMP) {
        fprintf(stderr, "-u and -T are mutually exclusive.\n");
        exit(EXIT_FAILURE);
      }
      mtrtype = IPPROTO_TCP;
      break;
    case 'P':
      remoteport = atoi(optarg);
      if (remoteport > 65535 || remoteport < 1) {
        fprintf(stderr, "Illegal port number.\n");
        exit(EXIT_FAILURE);
      }
      break;
    case 'z':
      timeout = atoi(optarg);
      timeout *= 1000000;
      break;
    case '4':
      af = AF_INET;
      break;
    case '6':
#ifdef ENABLE_IPV6
      af = AF_INET6;
      break;
#else
      fprintf( stderr, "IPv6 not enabled.\n" );
      break;
#endif
    }
  }

  if (DisplayMode == DisplayReport ||
      DisplayMode == DisplayTXT ||
      DisplayMode == DisplayXML ||
      DisplayMode == DisplayRaw ||
      DisplayMode == DisplayCSV)
    Interactive = 0;

  if (optind > argc - 1)
    return;

  Hostname = argv[optind++];

  if (argc > optind) {
    cpacketsize = atoi (argv[optind]);
  }
}


void parse_mtr_options (char *string)
{
  int argc;
  char *argv[128], *p;

  if (!string) return;

  argv[0] = "mtr";
  argc = 1;
  p = strtok (string, " \t");
  while (p != NULL && ((size_t) argc < (sizeof(argv)/sizeof(argv[0])))) {
    argv[argc++] = p;
    p = strtok (NULL, " \t");
  }
  if (p != NULL) {
    fprintf (stderr, "Warning: extra arguments ignored: %s", p);
  }

  parse_arg (argc, argv);
  optind = 0;
}


int main(int argc, char **argv) 
{
  ip_t *            traddr;
  struct hostent *  host                = NULL;
  int               net_preopen_result;
#ifdef ENABLE_IPV6
  struct addrinfo       hints, *res;
  int                   error;
  struct hostent        trhost;
  char *                alptr[2];
  struct sockaddr_in *  sa4;
  struct sockaddr_in6 * sa6;
#endif

  /*  Get the raw sockets first thing, so we can drop to user euid immediately  */

  if ( ( net_preopen_result = net_preopen () ) ) {
    fprintf( stderr, "mtr: unable to get raw sockets.\n" );
    exit( EXIT_FAILURE );
  }

  /*  Now drop to user permissions  */
  if (setgid(getgid()) || setuid(getuid())) {
    fprintf (stderr, "mtr: Unable to drop permissions.\n");
    exit(1);
  }

  /*  Double check, just in case  */
  if ((geteuid() != getuid()) || (getegid() != getgid())) {
    fprintf (stderr, "mtr: Unable to drop permissions.\n");
    exit(1);
  }

  /* reset the random seed */
  srand (getpid());
  
  display_detect(&argc, &argv);

  /* The field options are now in a static array all together, 
     but that requires a run-time initialization. -- REW */
  init_fld_options ();

  parse_mtr_options (getenv ("MTR_OPTIONS"));

  parse_arg (argc, argv);

  /* Now that we know mtrtype we can select which socket to use */
  if (net_selectsocket() != 0) {
    fprintf( stderr, "mtr: Couldn't determine raw socket type.\n" );
    exit( EXIT_FAILURE );
  }

  if (PrintVersion) {
    printf ("mtr " VERSION "\n");
    exit(0);
  }

  if (PrintHelp) {
    printf("usage: %s [-hvrwctglspniuT46] [--help] [--version] [--report]\n"
	   "\t\t[--report-wide] [--report-cycles=COUNT] [--curses] [--gtk]\n"
           "\t\t[--raw] [--split] [--mpls] [--no-dns] [--address interface]\n" /* BL */
           "\t\t[--psize=bytes/-s bytes]\n"            /* ok */
           "\t\t[--report-wide|-w] [-u|-T] [--port=PORT] [--timeout=SECONDS]\n"            /* rew */
	   "\t\t[--interval=SECONDS] HOSTNAME [PACKETSIZE]\n", argv[0]);
    exit(0);
  }

  if (Hostname == NULL) Hostname = "localhost";

  if (gethostname(LocalHostname, sizeof(LocalHostname))) {
	strcpy(LocalHostname, "UNKNOWNHOST");
  }

  if (net_preopen_result != 0) {
    fprintf(stderr, "mtr: Unable to get raw socket.  (Executable not suid?)\n");
    exit(1);
  }

#ifdef ENABLE_IPV6
  /* gethostbyname2() is deprecated so we'll use getaddrinfo() instead. */
  bzero( &hints, sizeof hints );
  hints.ai_family = af;
  hints.ai_socktype = SOCK_DGRAM;
  error = getaddrinfo( Hostname, NULL, &hints, &res );
  if ( error ) {
    if (error == EAI_SYSTEM)
       perror ("Failed to resolve host");
    else
       fprintf (stderr, "Failed to resolve host: %s\n", gai_strerror(error));
    exit( EXIT_FAILURE );
  }
  /* Convert the first addrinfo into a hostent. */
  host = &trhost;
  bzero( host, sizeof trhost );
  host->h_name = res->ai_canonname;
  host->h_aliases = NULL;
  host->h_addrtype = res->ai_family;
  af = res->ai_family;
  host->h_length = res->ai_addrlen;
  host->h_addr_list = alptr;
  switch ( af ) {
  case AF_INET:
    sa4 = (struct sockaddr_in *) res->ai_addr;
    alptr[0] = (void *) &(sa4->sin_addr);
    break;
  case AF_INET6:
    sa6 = (struct sockaddr_in6 *) res->ai_addr;
    alptr[0] = (void *) &(sa6->sin6_addr);
    break;
  default:
    fprintf( stderr, "mtr unknown address type\n" );
    exit( EXIT_FAILURE );
  }
  alptr[1] = NULL;
#else
    host = gethostbyname(Hostname);
  if (host == NULL) {
    herror("mtr gethostbyname");
    exit(1);
  }
  af = host->h_addrtype;
#endif

  traddr = (ip_t *) host->h_addr;
  
  if (net_open(host) != 0) {
	fprintf(stderr, "mtr: Unable to start net module.\n");
        exit(1);
      }

  if (net_set_interfaceaddress (InterfaceAddress) != 0) {
    fprintf( stderr, "mtr: Couldn't set interface address.\n" ); 
    exit( EXIT_FAILURE ); 
  }

  display_open();
  dns_open();

  display_mode = 0;
  display_loop();

  net_end_transit();
  display_close();
  net_close();

  return 0;
}


